docs/design/secrets.md

```diff
@@ -0,0 +1,194 @@
+# Secret Distribution
+
+## Abstract
+
+A proposal for the distribution of secrets (passwords, keys, etc) to containers inside Kubernetes
+using a custom volume type.
+
+## Motivation
+
+Secrets are needed in containers to access internal resources like the Kubernetes master or
+external resources such as git repositories, databases, etc. 
+
+A goal of this design is to eliminate or minimize the modifications to containers in order to
+access secrets. Secrets should be placed where the container expects them to be.
+
+## Constraints and Assumptions
+
+*  This design does not prescribe a method for storing secrets; storage of secrets should be
+   pluggable to accomodate different use-cases
+*  Encryption of secret data and node security are orthogonal concerns
+*  It is assumed that node and master are secure and that compromising their security could also
+   compromise secrets:
+   *  If a node is compromised, only the secrets for the containers scheduled on it should be
+      exposed
+   *  If the master is compromised, all secrets in the cluster may be exposed
+
+## Use Cases
+
+1.  As a user, I want to store secret artifacts for my applications and consume them securely in
+    containers, so that I can keep the configuration for my applications separate from the images
+    that use them:
+    1.  As a cluster operator, I want to allow a pod to access the Kubernetes master using a custom
+        `.kubeconfig` file, so that I can securely reach the master
+    2.  As a cluster operator, I want to allow a pod to access a Docker registry using credentials
+        from a `.dockercfg` file, so that containers can push images
+    3.  As a cluster operator, I want to allow a pod to access a git repository using SSH keys,
+        so that I can push and fetch to and from the repository
+2.  As a user, I want to allow containers to consume supplemental information about services such
+    as username and password which should be kept secret, so that I can share secrets about a
+    service amongst the containers in my application securely
+
```

Comment:
> somehow make sure that other pods that should not have access to that secret do use it for their images via some sort of race condition.

@erictune confirm that you meant "don't" instead of "do" here?